### ISSUE TITLE
[CI] Skip RISC-V compliance test or short tests

### ISSUE THREAD

**Issue by eunchan** [lowRISC/opentitan#1981]
It would be good if CI can skip RISC-V compliance test if no vendored-in ibex or rv_ibex_core is changed in the PR. The RISC-V Compliance test takes majority of the entire CI run time.

**Comment by sriyerg** [comment]
[Reposting the from the Slack #software channel]
If we enable the RISC-V compliance test in our DV nightly regressions, is there still a need to run it in CI?

**Comment by GregAC** [comment]
RTL changes other than Ibex can break riscv-compliance, in particular anything around tilelink or memory (I believe we had an issue with this in the past). So I'd prefer not to run it on Ibex vendoring only.

It's also possible SW changes would break it, though that's less of a concern as such things probably don't mean a SW bug but rather the riscv-compliance setup needing to adapt to the SW changes.

I think a nightly run could be a good compromise provided failures of it are quickly flagged up so we can get it fixed.

I'm going to see what I can do about speeding it up today. The run-time is oddly variable but I'm not sure there's much we can do about that (presumably to do with what physical system azure ends up running it on).

**Comment by GregAC** [comment]
@gkelly replying to your slack question here to avoid one discussion across multiple channels

**Comment by GregAC** [comment]
You can see the results of my experiments in parallelising the runs in #1987.

I've split it up into two jobs, one to run the large suite (rv32i) and the one to run the rest. I've also tried using -j2 on make which seems to work well. This gives around 8 minutes to complete everything (this is the riscv-compliance run alone ignoring the rest of the CI jobs) so provided things get scheduled so things run in parallel the riscv-compliance suite will finish before the FPGA build.

With that performance I think it's reasonable to include this in normal CI. I need to tidy up my PR (plus incorporate some minor tweaks from review feedback from the original PR) but that should be good to go in some time today.

**Comment by GregAC** [comment]
The one issue with parallel make is the log output is somewhat confusing, I'll see if I can alter the pytest logging output to give an indication of the test binary.

**Comment by imphil** [comment]
I see a roughly 4 minutes increase in CI latency, which equals to roughly a 20 percent increase.

So it's not for free, but let's give it a try and see how it goes. I expect the dominating factor to be the FPGA build in the near future as we grow our design, so we'll need to revisit questions and expectations around that anyways. (We're also not running into capacity issues with parallel builders most of the time currently, so running things in parallel shouldn't increase queuing for other jobs too much.)

**Comment by GregAC** [comment]
I'm am slightly puzzled it causes a latency increase. The FPGA build job takes longer than running the tests now with the latest changes and they should be able to start at the same time (it's possible the FPGA build job gets the single agent capable of running verilator/vivado builds before the earlgrey verilator is built but the verilator build generally seems to win that).

If the agent pool that can run the riscv compliance tests is fully utilised whilst the FPGA build starts then that will increase latency but in general we'd expect that to have capacity where the FPGA/verilator build agent is the bottleneck.

Latest PR incorporating some minor tweaks plus the parallelisation is #1990

**Comment by imphil** [comment]
Upon a closer look, the latency in an non-queued case shouldn't be impacted, there's a small margin between the FPGA job (12 min) and the build verilated sim + run verilator jobs (2:30 + 7:45). (https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=11202&view=results)

However, we are now more sensitive to queuing in the Default builder, which is needed for the FPGA job and the verilated sim build job.

But that's nothing we can solve here, the scope and approach of your solution is the right one IMO. Thanks for following up on all of that!

**Comment by eunchan** [comment]
Thank you guys for fast feedbacks. Let me close this issue as Greg followed with a PR.
